### ISSUE TITLE
grid_sample fails if "input" and "grid" have different types

### ISSUE THREAD

**Issue by madshi** [pytorch/pytorch#94733]
### 🐛 grid_sample needlessly requires "input" and "grid" to have the same type

I've found that when calling "grid_sample" with a float16 "input" and float32 "grid" parameter combo, pytorch errors out with:

RuntimeError: expected scalar type Half but found Float

Now this is easy to workaround, I can just use float32 for both parameters. However, this is undesired because it means that every time I call grid_sample, my "input" image has to be converted to float32 first, and then the grid_sample output has to be converted back to float16. Which is superfluous work which serves no purpose. It slows down both training and inference.

So instead I've tried keeping the image as float16 and calculate a float16 "grid". However, that doesn't work out properly. Even if I do all my grid calculation math in float32 and only convert the grid to "half()" right before calling grid_sample, I still get serious image quality problems.

To sum up: grid_sample should officially allow "input" and "grid" to have different types. There's zero justification (IMHO) to require that these two parameters have the same type, because the 2 parameters are not added or multiplied together or anything like that.

P.S: I've seen issue # 89836 which seems related. However, I think it might not be the grid_sample code which is causing inaccuracies with a float16 grid. I think probably float16 is simply not high resolution enough to produce accurate results. But anyway, even if there were no image quality issues with a float16 grid, I'd still argue that grid_sample should allow using different types for "grid" and "input".

### Versions

PyTorch 1.13.1

cc @albanD @mruberry @jbschlosser @walterddr @saketh-are

**Comment by drisspg** [comment]
Hey @madshi I am going to close this Issue in favor of this one: #94816

**Comment by madshi** [comment]
That's fine, thanks.
